src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkAppEngineFlexibleStaging.java

```diff
@@ -96,23 +96,26 @@ static String findRuntime(StageFlexibleConfiguration config)
   static void copyDockerContext(
       StageFlexibleConfiguration config, CopyService copyService, @Nullable String runtime)
       throws IOException, AppEngineException {
-    if (config.getDockerDirectory() != null && config.getDockerDirectory().exists()) {
+    File dockerDirectory = config.getDockerDirectory();
+    if (dockerDirectory != null && dockerDirectory.exists()) {
       if ("java".equals(runtime)) {
         log.warning(
             "WARNING: runtime 'java' detected, any docker configuration in "
-                + config.getDockerDirectory()
+                + dockerDirectory
                 + " will be ignored. If you wish to specify a docker configuration, please use "
                 + "'runtime: custom'.");
       } else {
         // Copy docker context to staging
-        if (!Files.isRegularFile(config.getDockerDirectory().toPath().resolve("Dockerfile"))) {
+        if (!Files.isRegularFile(dockerDirectory.toPath().resolve("Dockerfile"))) {
           throw new AppEngineException(
-              "Docker directory "
-                  + config.getDockerDirectory().toPath()
-                  + " does not contain Dockerfile.");
+              "Docker directory " + dockerDirectory.toPath() + " does not contain Dockerfile.");
         } else {
-          copyService.copyDirectory(
-              config.getDockerDirectory().toPath(), config.getStagingDirectory().toPath());
+          File stagingDirectory = config.getStagingDirectory();
```

Comment:
Can we enforce non-nullability on the interface `StageFlexibleConfiguration` directly? Since getStagingDirectory cannot be null.

Reply:
We definitely should take a deeper look at this and other things that are currently marked Nullable. We may be able to do some refactoring to guarantee non-nullability. Here we'd need to shift from using setter methods to a constructor or perhaps a builder.